lib/diagnostics_channel.js

```diff
@@ -0,0 +1,143 @@
+'use strict';
+
+const {
+  ArrayPrototypeIncludes,
+  ArrayPrototypePush,
+  FunctionPrototypeBind,
+  ObjectCreate,
+  ObjectSetPrototypeOf,
+} = primordials;
+
+const {
+  codes: {
+    ERR_INVALID_ARG_TYPE,
+  }
+} = require('internal/errors');
+
+// TODO(qard): should there be a C++ publisher interface?
+class ActivePublisher {
+  subscribe(subscription) {
+    if (typeof subscription === 'function') {
+      ArrayPrototypePush(this._subscribers, subscription);
+      return;
+    }
+
+    if (typeof subscription === 'object' &&
+        typeof subscription.onMessage === 'function') {
+      let shouldListen = true;
+
+      if (typeof subscription.shouldListen === 'function') {
+        shouldListen = subscription.shouldListen(this.name);
+      }
+
+      if (shouldListen) {
+        const fn = FunctionPrototypeBind(subscription.onMessage, subscription);
+        ArrayPrototypePush(this._subscribers, fn);
+      }
+    }
+  }
+
+  shouldPublish() {
+    return true;
+  }
+
+  publish(data) {
+    for (var i = 0; i < this._subscribers.length; i++) {
+      this._subscribers[i](this.name, data);
+    }
+  }
+}
+
+class Publisher {
+  constructor(name) {
+    this._subscribers = undefined;
+    this.ifActive = undefined;
+    this.name = name;
+  }
+
+  subscribe(subscription) {
+    ObjectSetPrototypeOf(this, ActivePublisher.prototype);
```

Comment:
I find this tinkering with prototype quite hard to read. Is it possible to get the same performance by using a "normal" class?

Reply:
Nope. Not even close. 😬